Add `go-netaddrs` support to `retry_join` (#18745)
hashicorp/nomad: website/content/docs/configuration/server_join.mdx

## `server_join` Parameters

- `retry_join` `(array<string>: [])` - Specifies a list of Nomad server
  addresses and [Cloud Auto-join](#cloud-auto-join) configurations that are
  joined as cluster members. This is similar to [`start_join`](#start_join),
  but join attempts are retried up to [retry_max](#retry_max) times.
  Further, `retry_join` is available to both Nomad servers and clients, while
  `start_join` is only defined for Nomad servers. This is useful for cases where
  we know the address will become available eventually. Use `retry_join` with an
  array as a replacement for `start_join`, **do not use both options**.

  Server addresses must conform to the [server address format](#server-address-format).

>   ```

Reviewer: ```suggestion
  ```hcl
```

  server_join {
    retry_join = [ "1.1.1.1", "2.2.2.2" ]
  }
  ```

  Auto-join configurations must conform to the [Cloud Auto-join format](#cloud-auto-join).
    

>   Cloud-Auto join using `go-discover` to join an AWS EC2 cluster.
>   ```

Reviewer: ```suggestion
  Cloud-Auto join using `go-discover` to join an AWS EC2 cluster.
  
  ```hcl
```

  server_join {
    retry_join = [ "provider=aws tag_key=..." ]
  }
  ```

  Cloud-Auto join using `go-netaddrs` to join a Hetzner cluster of servers
  labeled with `nomad-server=true` using the `hcloud` CLI.

  ```
  server_join {
    retry_join = [ "exec=hcloud server list -o noheader -o columns=ipv4 -l nomad-server=true | tr '\n' ' '" ]
  }
  ```

  Server addresses and Cloud Auto-join configurations may be used together.
  This is a mixed configuration containing a server address and Cloud Auto-join

>   configuration.
>   ```

Reviewer: ```suggestion
  configuration.
  
  ```hcl
```

nomad-01.company.local => nomad-01.company.local:4648
```

### `go-discover` format

> As of Nomad 0.8.4, `retry_join` accepts a unified interface using the

Reviewer: I have somewhere on a TODO list to remove all these ancient "as of Nomad 0.8..." references :grinning: 

```suggestion
`retry_join` accepts a unified interface using the
```

[go-discover](https://github.com/hashicorp/go-discover) library for doing
automated cluster joining using cloud provider metadata. See [Cloud
Auto-join](#cloud-auto-join) for more information.

```
"provider=aws tag_key=..." => 1.2.3.4:4648
```

### `go-netaddrs` format

> As of Nomad 1.x.x (**TODO**: What version would this land in?) `retry_join` accepts

Author: Calling out this TODO for a response.

Reviewer: We have versioned docs these days, so this would only appear in the major version in which the feature appears (ex. the 1.7.x docs). If we can manage to land this PR in 1.7.0, which seems likely, I think we can remove this clause entirely. Let's leave it out and if we need to adjust it I can come back in later to tweak it.

Author: Excellent. That sounds much better.

Discovery requires a [GCE Service
Account](https://cloud.google.com/compute/docs/access/service-accounts).
Credentials are searched using the following paths, in order of precedence.

### `go-netaddrs` Configurations

`go-netaddrs` configurations are prefixed with `exec=`.

What follows the `exec=` prefix may be any executable program and its arguments.
Commands run by `go-netaddrs` must return a list of space-delimited IPv4 or IPv6
addresses and exit with code `0` on success and non-zero on failure.

Executables may be referenced by absolute file system path or by any program

> name discoverable by `os/exec`'s [LookPath](https://pkg.go.dev/os/exec#LookPath)

Author: I have a feeling this may be too Go-specific for the audience. While `LookPath` is a nice system-independent way to talk about program path discovery, it's rooted in Go, which Nomad users may not be familiar with. 

Any thoughts on a system-independent way to discuss path discovery without being go-specific?

Reviewer: That's something worth considering, for sure (we run into this with our `template` docs and it's unfortunate). Here I think we can get away with a narrow description of the behavior and then refer to the Go docs:

> You can refer to the executable by its absolute file system path, or by the executable's name if it can be found on the Nomad agent's `PATH`. See [executables in the current directory](https://pkg.go.dev/os/exec#hdr-Executables_in_the_current_directory) for more details on the lookup behavior.

function.

> Example `go-netaddrs` configuration. 
> ```

Reviewer: ```suggestion
Example `go-netaddrs` configuration. 

```hcl
```

server_join {
  retry_join = [ "exec=hcloud server list -o noheader -o columns=ipv4 -l nomad-server=true | tr '\n' ' '" ]
}
```

> Here,  `hcloud`'s newline-delimited output has been re-formatted as space-
> deliminted by piping its output into `tr`.
> 
> Output of an exmaple `go-netaddrs` executable

Reviewer: ```suggestion
Here, `hcloud`'s newline-delimited output has been re-formatted as space-
delimited by piping its output into `tr`.

Output of an example `go-netaddrs` executable
```
